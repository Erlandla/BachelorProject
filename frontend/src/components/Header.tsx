import React, { useContext } from 'react'
import { Link } from 'react-router-dom'
import MenuIcon from '@mui/icons-material/Menu'
import CloseIcon from '@mui/icons-material/Close'
import Button from '@mui/material/Button'
import Menu from '@mui/material/Menu'
import PopupState, { bindTrigger, bindMenu } from 'material-ui-popup-state'
import { ProblemContext } from '../globalState/ProblemContext'

function Header() {
  const { user } = useContext(ProblemContext)

  return (
    <div className="h-[3.25rem] flex flex-row justify-between bg-white sticky top-0 z-50 drop-shadow-xl px-10 sm:px-20">
      <Link to="/Hjem" className="text-3xl text-left w-48 mt-2">
        <p data-cy = "nestaLogo" className="">Graph Nesta</p>
      </Link>

      <section className="text-based">
        <PopupState variant="popover" popupId="demo-popup-menu">
          {(popupState) => (
            <React.Fragment>
              <Button
                data-cy="hamburgerMenu"
                data-testid="hamburgerMenu"
                {...bindTrigger(popupState)}
              >
                {popupState.isOpen ? (
                  <CloseIcon
                    onClick={popupState.close}
                    sx={{ fontSize: '2.5rem', cursor: 'pointer', color: '#0D264A' }}
                  />
                ) : (
                  <MenuIcon
                    onClick={popupState.close}
                    sx={{ fontSize: '2.5rem', cursor: 'pointer', color: '#0D264A' }}
                  />
                )}
              </Button>
              <Menu {...bindMenu(popupState)}>
                {user.isLoggedIn ? (
                  <Link to="/minProfil" onClick={popupState.close}>
                    <div
                      data-cy="myProfile"
                      className="text-left pl-4 pr-4 h-10 hover:bg-menuHover flex items-centere"
                    >
                      Min profil
                    </div>
                  </Link>
                ) : (
                  <Link to="/LoggInn" onClick={popupState.close}>
                    <div
                      data-cy="loggin"
                      className="text-left pl-4 pr-4 h-10 hover:bg-menuHover flex items-center"
                    >
                      Logg inn
                    </div>
                  </Link>
                )}
                <Link to="/MineProblem" onClick={popupState.close}>
                  <div
                    data-cy="myProblems"
                    className="text-left pl-4 pr-4 h-10 hover:bg-menuHover flex items-center"
                  >
                    Mine problem
                  </div>
                </Link>
                <Link to="/NyttProblem" onClick={popupState.close}>
                  <div className="text-left pl-4 pr-4 h-10 hover:bg-menuHover flex items-center">
                    Nytt problem
                  </div>
                </Link>
                <Link
                  to="/Søk"
                  onClick={() => {
                    popupState.close()
                    localStorage.setItem('searchFilter', '0')
                  }}
                >
                  <div className="text-left pl-4 pr-4 h-10 hover:bg-menuHover flex items-center">
                    Søk
                  </div>
                </Link>
                {user.isAdmin.toString() === 'true' ? (
                  <div className="flex flex-col">
                    <Link to="/GodkjennProblem" onClick={popupState.close}>
                      <div className="text-left pl-4 pr-4 h-10 hover:bg-menuHover flex items-center">
                        Godkjenn problem
                      </div>
                    </Link>
                    <Link to="/RegistrerBruker" onClick={popupState.close}>
                      <div
                        data-cy="newUser"
                        className="text-left pl-4 pr-4 h-10 hover:bg-menuHover flex items-center"
                      >
                        Ny bruker
                      </div>
                    </Link>
                  </div>
                ) : null}
              </Menu>
            </React.Fragment>
          )}
        </PopupState>
      </section>
    </div>
  )
}

export default Header
