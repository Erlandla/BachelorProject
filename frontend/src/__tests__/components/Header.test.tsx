import React from "react";
import { render, screen, cleanup } from "@testing-library/react";
import Header from "../../components/Header";
import { MemoryRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import renderer from "react-test-renderer";

afterEach(() => {
    cleanup();
})

describe("Correctly renders elements", () => {
    it("'Graph Nesta' renders", () => {
        render(
            <MemoryRouter>
                <Header />
            </MemoryRouter>
        );
        expect(screen.getByText("Graph Nesta")).toBeInTheDocument();
    });

    it("Hamburger menu appears, is clickable, and reveals extra content", () => {
        render(
            <MemoryRouter>
                <Header />
            </MemoryRouter>
        );
        userEvent.click(screen.getByTestId("hamburgerMenu"));
        expect(screen.getByText("Logg inn")).toBeInTheDocument();
        expect(screen.getByText("Mine problem")).toBeInTheDocument();
        expect(screen.getByText("Nytt problem")).toBeInTheDocument();
        expect(screen.getByText("SÃ¸k")).toBeInTheDocument();
    });

});

it("Matches snapshot", () => {
    const snapshot = renderer.create(
        <MemoryRouter>
            <Header />
        </MemoryRouter>
    ).toJSON();
    expect(snapshot).toMatchSnapshot();
})