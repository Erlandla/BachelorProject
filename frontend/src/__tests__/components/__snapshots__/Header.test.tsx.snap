// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<div
  className="h-[3.25rem] flex flex-row justify-between bg-white sticky top-0 z-50 drop-shadow-xl px-10 sm:px-20"
>
  <a
    className="text-3xl text-left w-48 mt-2"
    href="/Hjem"
    onClick={[Function]}
  >
    <p
      className=""
    >
      Graph Nesta
    </p>
  </a>
  <section
    className="text-based"
  >
    <button
      aria-haspopup={true}
      className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      data-cy="hamburgerMenu"
      data-testid="hamburgerMenu"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i7v600-MuiSvgIcon-root"
        data-testid="MenuIcon"
        focusable="false"
        onClick={[Function]}
        viewBox="0 0 24 24"
      >
        <path
          d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
        />
      </svg>
    </button>
  </section>
</div>
`;
